### ISSUE TITLE
Bookloupe error message

### ISSUE THREAD

**Issue by okrick** [DistributedProofreaders/guiguts#720]
**Describe the bug**
13:15:56: Guiguts 1.3.0: If you have any problems, please include any error messages that appear here with your bug report.
13:19:23: Tk::Error: Tk::Toplevel=HASH(0x8f94200) is not a widget at C:/DP/guiguts/lib/Guiguts/Utilities.pm line 1599.
 Tk callback for .toplevel1.frame.frame.button1
 <ButtonRelease-1>
 (command bound to event)
13:19:23: Tk::Error: Usage $widget->update(...)
Tk::Frame=HASH(0x8f7c760) is not a Tk object at C:/DP/guiguts/lib/Guiguts/ErrorCheck.pm line 388.
 Guiguts::ErrorCheck::errorcheckpop_up at C:/DP/guiguts/lib/Guiguts/ErrorCheck.pm line 388
 Guiguts::KeyBindings::__ANON__ at C:/DP/guiguts/lib/Guiguts/KeyBindings.pm line 40
 <Key-F6>
 (command bound to event)

**To Reproduce**
1. Open Guiguts
2. Press F6
3. Hide all except for Short lines
4. Close the Bookloupe window (either by clicking the x in the upper right corner or Alt+F4)
5. See error message

**Expected behavior**
I didn't expect to see the error message.  

Also, I noticed the Perl interpreter took less than 1% of the CPU resources until Bookloupe was started then it went to 24% of CPU and stayed that way as long as the Bookloupe window was open.  For comparison, the WF window does the same temporarily then after the processing was done Perl's drain on the CPU dropped to less than 1% with the WF window still open.  I don't know if this has any bearing on anything but thought I'd mention it.  The reason I was checking the resources was that I was seeing a slow down in GuiGuts and other programs.

**Environment**
* Guiguts version:  1.3.0
* OS and version: Windows 10 Home 20H2
* Perl version: Strawberry Perl (64-bit) 5.32.1.1-64bit
* Perl distribution: Strawberry Perl (64-bit) 5.32.1.1-64bit

**Additional context**
[13949.zip](https://github.com/DistributedProofreaders/guiguts/files/6338931/13949.zip)

I closed and reopened GuiGuts and still get the same error message.

**Comment by windymilla** [comment]
Here's the explanation of what is happening (and would have happened in old versions too under the same circumstances). As you will see, it's a sad confluence of circumstances:
1. When you run a check tool, like bookloupe, the program inserts marks where all the errors are to assist you in jumping to them from the error list dialog.
2. Because your file is French, it has many non-ASCII characters and unexpected words, leading to over 27000 warnings, so 27000 marks are added to the file.
3. Similar internal marks are also used to flag page boundaries.
4. Your file does not have any page boundaries (never happens with a normal project coming from the rounds)
5. There is a routine that is called every fifth of a second to update the status bar and a few other things - this routine does not normally take long to complete
6. One thing the routine updates is the current page number as shown in the status bar.
7. To determine the page number, it looks backwards through the marks trying to find one which flags a page boundary. If it fails, it looks forwards.
8. Normally, even if there were 27000 warning marks, there would only be a few (probably less than 100) on a page before it finds a page marker, so it doesn't take long to find one.
9. In your file, it ends up searching through all 27000 marks looking for a page mark.
10. This takes almost (or maybe more than) a fifth of a second, so as soon as it is finished, the routine is called again, and again, and again, etc., so there's virtually no time for anything else to happen, which explains the very sluggish behaviour.
11. The error message is caused because (due to this sluggishness) you actually end up crossing off the dialog before the routine that updates it has finished, so it is trying update a deleted window.

I have a fix that involves checking to see if the line and column have changed since the last call - if not, then there's no need to try to find the page number again. In the meantime, here's a minor edit you can make, @okrick, to enable you to work with the file - it will make the status bar updating slightly delayed (may be 1 second to update) but the rest of the program should be responsive.
In StatusBar.pm, about line 15, is the start of the update_indicators subroutine:
```
sub update_indicators {
```
Change it by adding a line above and below it as follows
```
my $update_skipper = 0;
sub update_indicators {
    return if ++$update_skipper % 5;
```
Essentially this means it will only do an update of the status bar once in every five times it is called, i.e. about once a second - you may notice a slight pause every second as you work. You can change the 5 to a larger or smaller number - smaller makes it update more quickly, but runs the risk of making the whole program sluggish, large has the opposite results: slower status update, but more responsive program.

**Comment by windymilla** [comment]
Alternative workaround is to not edit the code, but use File->Project->Guess Page Markers to insert some page marks so it doesn't waste lots of time searching. 
The exact numbers don't really matter. You can assume about 100 lines per page, so tell it that it has 270 pages, that page 25 starts at line 2500, and that page 260 starts at line 26000.

**Comment by okrick** [comment]
Thank you.  I'll use the guess page numbers fix in the future when working the errata fixes.  

This French series was one where the Bookloupe was more helpful than the pptext text checks.  I could at least hide the options that would be valid only in English.  And this series did not have a previous HTML so I had to create one with Guiguts.

**Comment by windymilla** [comment]
@okrick - my fix seems to work, so will hopefully be in 1.3.1 when it's released, and you won't have the problem again. In the meantime, if you notice the same issue, try the Guess Page Numbers workaround.
